File: apps/desktop/src/hotkeys.ts
import type { HotkeyBinding } from '@clipped/shared';
import { invoke } from '@tauri-apps/api/core';
import { useCallback, useEffect, useState } from 'react';

import { asProblem, type LibraryProblem, type LibraryRead } from './library';
import { saveSettings } from './settings';

/**
 * Reading the global hotkeys from the window.
 *
 * The window registers none of them. `RegisterHotKey` gives a combination to
 * exactly one process, and that process is the recorder: it is the one that
 * starts at login and outlives every window, and the one that can act on a press
 * (`docs/adr/0009-the-recorder-registers-global-hotkeys.md`, `docs/hotkeys.md`).
 *
 * That has a consequence this module exists for. The recorder finds out that
 * Discord already owns `Ctrl+F10` at the moment it registers, which may be days
 * before anybody opens this window — so a conflict cannot arrive as an event,
 * and a window that waited for one would show a clean list for ever. It is asked
 * for instead, whenever the screen that draws it is opened.
 *
 * # Three answers, never two
 *
 * The same three {@link LibraryRead} carries, and for the same reason: an empty
 * list is not the same as a recorder that could not be asked, and drawing the
 * second as the first would say "no hotkey has a problem" about a recorder that
 * was never reached (AGENTS.md section 27).
 */

/** Asks the recorder where every hotkey stands. */
export async function readHotkeys(): Promise<readonly HotkeyBinding[]> {
  return invoke<HotkeyBinding[]>('recorder_hotkeys');
}

/**
 * What the window says about a hotkey list it could not read.
 *
 * `unknown_command` is the one worth its own sentence. A recorder that started
 * before this window was installed registers no hotkeys at all and has no
 * `get_hotkeys` command, and "your hotkeys could not be read" would send
 * somebody looking for a fault in their keyboard rather than restarting Clipped
 * (AGENTS.md section 45).
 */
export function describeHotkeyProblem(problem: LibraryProblem): string {
  switch (problem.code) {
    case 'no_recorder_configured':
    case 'recorder_unreachable':
      return `Clipped could not ask the recorder about your hotkeys, so nothing here is known. ${problem.message}`;
    case 'unknown_command':
      return 'The recorder that is running is older than this window and registers no hotkeys at all, so none of them does anything. Restarting Clipped starts the recorder that came with it.';
    default:
      return problem.message;
  }
}

/**
 * What one row says about whether that key does anything.
 *
 * Three states rather than two, because there are three ways a hotkey ends up
 * doing nothing and the useful action differs in each:
 *
 * - **unbound** — nothing is pointed at it. Not a fault; five of the seven
 *   actions start this way.
 * - **conflict** — Windows would not give Clipped the combination. The sentence
 *   is the recorder's own and names who is likely to have it.
 * - **unavailable** — the combination is registered and nothing in this build
 *   performs the action, so pressing it reports itself rather than working.
 */
export type HotkeyCondition = 'working' | 'unbound' | 'conflict' | 'unavailable';

/** Which of the four a row is in. */
export function conditionOf(row: HotkeyBinding): HotkeyCondition {
  if (row.state.state === 'conflict') {
    return 'conflict';
  }
  if (!row.handled) {
    return 'unavailable';
  }
  return row.state.state === 'unbound' ? 'unbound' : 'working';
}

/**
 * What that condition says, in the words shown beside the row.
 *
 * The recorder's own sentence wherever there is one: only the process that asked
 * Windows knows who has the combination, and only it knows which milestone would
 * build the action. This window invents no wording for either.
 */
export function describeCondition(row: HotkeyBinding): string {
  switch (conditionOf(row)) {
    case 'conflict':
      return row.state.state === 'conflict' ? row.state.reason : '';
    case 'unavailable':
      return row.unavailable ?? '';
    case 'unbound':
      return 'Not bound to anything.';
    case 'working':
      return 'Registered, and this build performs it.';
  }
}

/** The hotkeys, read once when the screen that shows them is opened. */
export function useHotkeys(): LibraryRead<readonly HotkeyBinding[]> {
  const [read, setRead] = useState<LibraryRead<readonly HotkeyBinding[]>>({ state: 'reading' });

  useEffect(() => {
    let current = true;
    readHotkeys()
      .then((hotkeys) => {
        if (current) {
          setRead({ state: 'read', value: hotkeys });
        }
      })
      .catch((thrown: unknown) => {
        if (current) {
          setRead({ state: 'unread', problem: asProblem(thrown) });
        }
      });
    return () => {
      current = false;
    };
  }, []);

  return read;
}

/**
 * The name this window sends a hotkey under.
 *
 * `apps/recorder/src/settings.rs` prefixes every hotkey key with `hotkey_` so
 * that one `apply_settings` can carry a combination and a frame rate without
 * two vocabularies. The prefix is duplicated here rather than derived, because
 * the alternative is a protocol field whose only job is to spell a constant —
 * `apps/recorder/tests/ipc_protocol.rs` is where the two are held together.
 */
export const HOTKEY_SETTING_PREFIX = 'hotkey_';

/**
 * What a hotkey is spelled as when it is deliberately bound to nothing.
 *
 * `apps/recorder/src/settings.rs` calls this `UNBOUND`, and it is **not** the
 * same as clearing the setting. `null` is Reset: it stops the file saying
 * anything, so the action follows whatever Clipped ships with, today and after
 * the day that default changes. This word is the user saying "nothing", which
 * survives that change.
 *
 * The two are one keystroke apart in a settings screen and a year apart in
 * consequence, which is why the window has a control for each rather than one
 * labelled Clear.
 */
export const UNBOUND = 'none';

/** What a chord this window captured is called on the wire. */
export function hotkeySettingKey(action: string): string {
  return `${HOTKEY_SETTING_PREFIX}${action}`;
}

/**
 * The combination a key press names, or `null` when it names none yet.
 *
 * Windows binds a *modified* key, so a press of `Shift` alone is somebody part
 * way through a chord rather than a chord — returning `null` for it is what
 * lets a capture control wait instead of saving `Shift`.
 *
 * The spelling is `Hotkey`'s own (`Ctrl+Shift+F9`): modifiers in a fixed order
 * so the same chord is one string however it was typed, then the key. It has to
 * match, because the recorder parses this with `FromStr` and answers with
 * `Display` — a window that spelled it `Control+F9` would save a setting the
 * recorder refuses.
 */
export function chordOf(event: {
  readonly key: string;
  readonly ctrlKey: boolean;
  readonly shiftKey: boolean;
  readonly altKey: boolean;
}): string | null {
  const key = namedKey(event.key);
  if (key === null) {
    return null;
  }

  const parts: string[] = [];
  if (event.ctrlKey) {
    parts.push('Ctrl');
  }
  if (event.altKey) {
    parts.push('Alt');
  }
  if (event.shiftKey) {
    parts.push('Shift');
  }

  // A bare key is not a global hotkey. `RegisterHotKey` would take `F9` and
  // then every press of F9 in every application would belong to Clipped, which
  // is not a thing a user asked for by pressing one key in a settings screen.
  if (parts.length === 0) {
    return null;
  }

  parts.push(key);
  return parts.join('+');
}

/** The key half of a chord, or `null` when the press was a modifier alone. */
function namedKey(key: string): string | null {
  if (key === 'Control' || key === 'Shift' || key === 'Alt' || key === 'Meta') {
    return null;
  }
  if (/^F([1-9]|1\d|2[0-4])$/.test(key)) {
    return key;
  }
  if (/^[a-zA-Z0-9]$/.test(key)) {
    return key.toUpperCase();
  }
  return null;
}

/**
 * The action already holding `chord`, if this window can see one.
 *
 * "Where detectable" is the whole of what this claims, and issue #54's second
 * criterion says it that way for a reason. Two of Clipped's own actions on one
 * combination is a conflict this window can see in the list it already has, and
 * saving it would trade one working hotkey for another. A combination *another
 * application* owns is not visible from here at all — only the process that
 * calls `RegisterHotKey` learns that, which is the recorder, and it reports it
 * on the next read.
 */
export function actionHolding(
  rows: readonly HotkeyBinding[],
  chord: string,
  except: string,
): HotkeyBinding | null {
  return rows.find((row) => row.action !== except && row.hotkey === chord) ?? null;
}

/** Saving a combination, and what came back. */
export interface HotkeyEditor {
  /** The rows as they now stand, or `null` until the first read answers. */
  readonly rows: readonly HotkeyBinding[] | null;
  /** The action being saved, or `null` when nothing is in flight. */
  readonly pending: string | null;
  /** What went wrong with the last save, if anything did. */
  readonly problem: LibraryProblem | null;
  /** Binds `action` to `chord`, or unbinds it when `chord` is `null`. */
  readonly bind: (action: string, chord: string | null) => Promise<void>;
}

/**
 * Binding a combination from this window (issue #54).
 *
 * One save does the whole of it. `apply_settings` writes the combination into
 * the settings file *and* rebinds the running service — `apps/recorder/src/hotkeys.rs`
 * is where the second half happens, and it is why the first acceptance
 * criterion, "rebinding takes effect immediately", is a property of the
 * recorder rather than something this window arranges by asking twice
 * (issue #233, closed).
 *
 * What this window does have to do is **read the answer back**. The reply says
 * what the settings are, not what Windows made of them, and those differ
 * exactly when it matters: a combination another application owns saves
 * perfectly and then does not register. So a save is followed by a re-read, and
 * the row the user just changed shows the recorder's verdict rather than the
 * value that was sent.
 */
export function useHotkeyEditor(read: readonly HotkeyBinding[] | null): HotkeyEditor {
  const [edited, setEdited] = useState<readonly HotkeyBinding[] | null>(null);
  const [pending, setPending] = useState<string | null>(null);
  const [problem, setProblem] = useState<LibraryProblem | null>(null);

  const bind = useCallback(async (action: string, chord: string | null): Promise<void> => {
    setPending(action);
    setProblem(null);
    try {
      await saveSettings({ [hotkeySettingKey(action)]: chord });
      // The verdict, not the request. See above.
      setEdited(await readHotkeys());
    } catch (thrown: unknown) {
      setProblem(asProblem(thrown));
    } finally {
      setPending(null);
    }
  }, []);

  return { rows: edited ?? read, pending, problem, bind };
}

File: crates/ipc/src/hotkeys.rs
//! What the recorder's global hotkeys are doing, as the window reads it.
//!
//! # Why this is on the protocol at all
//!
//! `RegisterHotKey` gives a combination to exactly one process, and
//! [ADR 0009](../../../docs/adr/0009-the-recorder-registers-global-hotkeys.md)
//! made that process the recorder: it is the one that outlives every window, and
//! the one that can act on a press. So the window is not the process that finds
//! out that Discord already owns `Ctrl`+`F10` — the recorder is, at the moment it
//! registers, which may be days before anybody opens a window.
//!
//! A combination the user cannot have is exactly the failure a hotkey has: the
//! key does nothing and nothing says why. Writing it to the recorder's log
//! satisfies nobody, so `get_hotkeys` is the question the window asks and this is
//! the answer (AGENTS.md sections 27 and 45).
//!
//! # Why it is asked for rather than pushed
//!
//! Registration happens once, when the recorder starts. An event announcing a
//! conflict would be published before any window existed to hear it, and the
//! window that opened an hour later would show nothing — which is the state this
//! whole reply exists to prevent. A question has no such race: whenever the
//! window asks, the answer is the current one.
//!
//! # Why these types do not come from `clipped-hotkeys`
//!
//! This crate depends on no other crate of the workspace and may not start
//! (ADR 0002, `tests/integration/tests/workspace_layering.rs`): the desktop
//! application links it, and nothing of the recording engine may come with it.
//! So these are wire types, and the recorder maps its own
//! `clipped_hotkeys::Registration` onto them — the same relationship
//! [`RecorderStatus`](crate::RecorderStatus) has with a recording.

use serde::{Deserialize, Serialize};

/// One action, what it is bound to, and whether pressing it would do anything.
///
/// One row of the hotkey list a window draws. Every field answers a question
/// somebody looking at that row is asking, and the two that matter most are the
/// two a naive implementation leaves out: [`state`](Self::state), because a
/// binding Windows refused is not a binding, and
/// [`unavailable`](Self::unavailable), because a *registered* combination whose
/// action nothing performs is still a key that does nothing.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct HotkeyBinding {
    /// The action's stable name, such as `save_replay`.
    ///
    /// The same name the command it performs has where there is one, so a log
    /// line about `add_bookmark` means one thing whichever side wrote it.
    pub action: String,
    /// The action's name in the words a person reads, such as `Save replay`.
    ///
    /// Sent rather than derived, so that the list of actions has one home. A
    /// window that kept its own table of labels would be a second answer to
    /// "what is this action called", and would silently show nothing at all for
    /// an action a newer recorder had added.
    pub label: String,
    /// The combination, written as `Ctrl+F10`. Absent when the action is bound
    /// to nothing.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hotkey: Option<String>,
    /// What Windows said about it.
    pub state: HotkeyState,
    /// Whether anything in the recorder performs the action.
    ///
    /// A registered combination with no handler still does nothing when pressed.
    /// Drawing such a row as working would be the lie AGENTS.md section 27 is
    /// about, so this is a field rather than an inference from
    /// [`state`](Self::state).
    pub handled: bool,
    /// Why pressing it would do nothing, when that is the case.
    ///
    /// Present exactly when [`handled`](Self::handled) is `false`, and it is the
    /// recorder's own sentence: "Save replay is not in this build: a recording
    /// with a replay buffer arrives in M3 (issue #38)". Absent for an action the
    /// recorder performs.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unavailable: Option<String>,
}

/// What Windows said about one binding.
///
/// **Closed, and deliberately without a catch-all**, for the reason
/// [`RecorderStatus`](crate::RecorderStatus) has none (`docs/ipc.md`): a state a
/// client cannot read is one it would otherwise draw as something else, and the
/// something else here is "this hotkey works". An unreadable frame is the honest
/// outcome; an event a client cannot read is information it does without, and
/// this is not an event.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum HotkeyState {
    /// The action has no combination, so nothing was registered.
    Unbound,
    /// Windows accepted it, and presses are being delivered.
    Registered,
    /// Windows refused it, most often because another application owns it.
    Conflict {
        /// What to tell the user, in the recorder's own words: what failed, who
        /// is likely to have it, and what to do next (AGENTS.md section 45).
        reason: String,
    },
}

#[cfg(test)]
mod tests {
    use super::{HotkeyBinding, HotkeyState};

    #[test]
    fn an_unbound_action_leaves_the_combination_out_rather_than_sending_an_empty_one() {
        let row = HotkeyBinding {
            action: "open_overlay".to_owned(),
            label: "Open overlay".to_owned(),
            hotkey: None,
            state: HotkeyState::Unbound,
            handled: false,
            unavailable: Some("the in-game overlay arrives in M5 (issue #53)".to_owned()),
        };

        let json = serde_json::to_string(&row).expect("it serialises");
        assert_eq!(
            json,
            r#"{"action":"open_overlay","label":"Open overlay","state":{"state":"unbound"},"handled":false,"unavailable":"the in-game overlay arrives in M5 (issue #53)"}"#,
        );
        assert_eq!(
            serde_json::from_str::<HotkeyBinding>(&json).expect("and deserialises"),
            row,
        );
    }

    /// The row this whole type exists for: a hotkey the user cannot have,
    /// carrying the sentence that says why.
    #[test]
    fn a_refused_combination_carries_the_reason_across_the_wire() {
        let row = HotkeyBinding {
            action: "save_replay".to_owned(),
            label: "Save replay".to_owned(),
            hotkey: Some("Ctrl+F10".to_owned()),
            state: HotkeyState::Conflict {
                reason: "another application already uses it".to_owned(),
            },
            handled: false,
            unavailable: None,
        };

        let json = serde_json::to_string(&row).expect("it serialises");
        let back: HotkeyBinding = serde_json::from_str(&json).expect("and deserialises");
        assert_eq!(back, row);
        assert!(
            json.contains("another application already uses it"),
            "a conflict that arrives without its reason is a log line with extra steps: {json}",
        );
    }

    /// `docs/ipc.md`: an unknown *state* must make the frame unreadable rather
    /// than be smoothed into something plausible.
    #[test]
    fn a_state_this_build_has_never_heard_of_is_refused_rather_than_guessed() {
        let json = r#"{"action":"save_replay","label":"Save replay","state":{"state":"pending"},"handled":true}"#;

        assert!(
            serde_json::from_str::<HotkeyBinding>(json).is_err(),
            "a state this build cannot read must not parse as one it can",
        );
    }
}
